testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/ejb/remote/ClientClusterNodeSelectorTestCase.java

```diff
@@ -0,0 +1,219 @@
+                , BMRunnerUtil.computeBMScriptName(ClientClusterNodeSelectorTestCase.class.getAnnotation(BMScript.class).value())
+                , BMRunnerUtil.normaliseLoadDirectory(ClientClusterNodeSelectorTestCase.class.getAnnotation(BMScript.class)));
+
+        // trigger the static init of the correct properties file - this also depends on running in forkMode=always
+        JBossEJBProperties ejbProperties =
+                JBossEJBProperties.fromClassPath(ClientClusterNodeSelectorTestCase.class.getClassLoader()
+                        , PROPERTIES_FILE);
+        JBossEJBProperties.getContextManager().setGlobalDefault(ejbProperties);
+    }
+
+    @AfterClass
+    public static void afterClass() throws Exception {
+        BMUnitConfigState.popConfigurationState(ClientClusterNodeSelectorTestCase.class);
+    }
+
+    @Deployment(name = DEPLOYMENT_1, managed = false, testable = false)
+    @TargetsContainer(NODE_1)
+    public static Archive<?> createDeploymentForContainer1() {
+        return createDeployment(MODULE_NAME);
+    }
+
+    @Deployment(name = DEPLOYMENT_2, managed = false, testable = false)
+    @TargetsContainer(NODE_2)
+    public static Archive<?> createDeploymentForContainer2() {
+        return createDeployment(MODULE_NAME);
+    }
+
+    @Deployment(name = DEPLOYMENT_3, managed = false, testable = false)
+    @TargetsContainer(NODE_3)
+    public static Archive<?> createDeploymentForContainer3() {
+        return createDeployment(MODULE_NAME);
+    }
+
+    @Deployment(name = DEPLOYMENT_4, managed = false, testable = false)
+    @TargetsContainer(NODE_4)
+    public static Archive<?> createDeploymentForContainer4() {
+        return createDeployment(MODULE_NAME);
+    }
+
+    /**
+     * Instructs the ClusterNodeSelector on which node to call before each request and that checks that the response
+     * comes from that very node;
+     * @param baseURL1
+     * @throws Exception
+     */
+    @Test
+    @RunAsClient
+    public void test(
+            @ArquillianResource() @OperateOnDeployment(DEPLOYMENT_1) URL baseURL1
+    ) throws Exception {
+
+        final EJBClientContext.Builder ejbClientBuilder = new EJBClientContext.Builder();
+        ejbClientBuilder.addTransportProvider(new RemoteTransportProvider());
+        final EJBClientConnection.Builder connBuilder = new EJBClientConnection.Builder();
+        connBuilder.setDestination(URI.create(String.format("remote+%s", baseURL1)));
+        ejbClientBuilder.addClientConnection(connBuilder.build());
+        // ====================================================================================
+        // set the custom ClusterNodeSelector
+        // ====================================================================================
+        ejbClientBuilder.setClusterNodeSelector(new CustomClusterNodeSelector());
+        final EJBClientContext ejbCtx = ejbClientBuilder.build();
+        EJBClientContext.getContextManager().setThreadDefault(ejbCtx);
```

Comment:
I think this may also need to be cleaned up at the end of the test

Reply:
@stuartwdouglas because the ejb client relies heavily on static context, all ejb client tests need to be isolated by disabling reuseForks in the surefire test profile - so technically, thread context is destroyed at the end of the test.
@tommaso-borgato That said, it's a good idea to perform this cleanup anyway within the test, in the event that this thread isolation requirement goes away, either within a finally block within the test method, or an @After method.

Reply:
Thank you @pferraro : added clean up code